python-squid/src/main/java/org/sonar/python/metrics/MetricsVisitor.java

```diff
@@ -0,0 +1,170 @@
+  private Set<Integer> noSonar = new HashSet<>();
+  private boolean firstNonCommentSeen = false;
+
+  public MetricsVisitor(boolean ignoreHeaderComments) {
+    this.ignoreHeaderComments = ignoreHeaderComments;
+  }
+
+  @Override
+  public void visitElement(PsiElement element) {
+    // track the first meaningful element which is not a comment to handle ignoreHeaderComments
+    if (!((element instanceof PyFile) || (element instanceof PsiComment) || (element instanceof PsiWhiteSpace))) {
+      firstNonCommentSeen = true;
+    }
+
+    // track lines of code
+    if (element instanceof LeafPsiElement
+      && !(element instanceof PsiWhiteSpace)
+      && !(element instanceof PsiComment)
+      && element.getNode().getElementType() != PyTokenTypes.DOCSTRING) {
+      PythonTokenLocation location = new PythonTokenLocation(element);
+      for (int line = location.startLine(); line <= location.endLine(); line++) {
+        linesOfCode.add(line);
+      }
+    }
+
+    // track executable lines of code
+    if (element instanceof PyStatement) {
+      handlePyStatement(element);
+    }
+
+    // track lines of comments and no_sonar lines
+    if (element instanceof PsiComment) {
+      if (ignoreHeaderComments && !firstNonCommentSeen) {
+        firstNonCommentSeen = true;
+      } else {
+        handleComment(((PsiComment) element));
+      }
+    }
+
+    super.visitElement(element);
+  }
+
+  private void handlePyStatement(PsiElement element) {
+    if (!isDocString(element)) {
+      executableLines.add(new PythonTokenLocation(element).startLine());
+    }
+    if (element instanceof PyIfStatement) {
+      for (PyIfPart pyIfPart : ((PyIfStatement) element).getElifParts()) {
+        executableLines.add(new PythonTokenLocation(pyIfPart).startLine());
+      }
+    } else if (element instanceof PyTryExceptStatement) {
+      for (PyExceptPart pyExceptPart : ((PyTryExceptStatement) element).getExceptParts()) {
+        executableLines.add(new PythonTokenLocation(pyExceptPart).startLine());
+      }
+    }
+  }
+
+  private void handleComment(PsiComment comment) {
+    String[] commentLines = getCommentContent(comment.getText())
+      .split("(\r)?\n|\r", -1);
+    int line = new PythonTokenLocation(comment).startLine();
+
+    for (String commentLine : commentLines) {
+      if (commentLine.contains("NOSONAR")) {
+        noSonar.add(line);
+      } else if (!isBlankComment(commentLine) && !noSonar.contains(line)) {
+        linesOfComments.add(line);
+      }
+      line++;
+    }
+  }
+
+  @Override
+  public void visitPyStringLiteralExpression(PyStringLiteralExpression node) {
+    if (node.isDocString()) {
+      PythonTokenLocation location = new PythonTokenLocation(node);
+      for (int line = location.startLine(); line <= location.endLine(); line++) {
+        linesOfComments.add(line);
+      }
+    }
+    super.visitPyStringLiteralExpression(node);
+  }
+
+  private static boolean isDocString(PsiElement element) {
```

Comment:
I would change the type of `element` to `PyStatement` so that the body of the method makes more sense.

Reply:
@pynicolas https://github.com/SonarSource/sonar-python/pull/254/commits/089dc2c93c9fa732f44c09ce526d0c697997f5c4.